Architecture Diagrams (#201)
mulesoft/docs-gateway: gateway/1.4/modules/ROOT/pages/flex-architecture-uc-overview.adoc

= Integration Use Cases
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> The following integration use cases are determined by the ownership of the API Provider and API Consumer. Each use case shows physical patterns that demonstrate the necessary services in the system. 

Reviewer: Just checking whether "Provider" and "Consumer" should be lower-case instead of upper-case. You use lower case for them later on.

For the following use case diagrams, Flex Gateway is running in Connected Mode. However, you can also run Flex Gateway in Local Mode or use a combination of Connected Mode and Local Mode for separate runtimes in the same or in different environments.

The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks that have the appropriate network controls applied.

Flex Gateway supports four use cases:

* xref:flex-architecture-uc1.adoc[]
* xref:flex-architecture-uc2.adoc[] 
* xref:flex-architecture-uc3.adoc[] 
* xref:flex-architecture-uc4.adoc[] 

image:ad-intergration-use-cases.png[The four use cases are linked to products that they can support.]

Each use case supports different integration patterns that vary depending on the location of APIs relative to the network in which Flex Gateway is located. You can use different use case patterns, depending on the different API owners and networks in your system. 

> In each use case, there are three different classifications of networks. The classifications describe the networks relation to your organization and the network Flex Gateway is in:

Reviewer: "the networks relation to your organization" - do you mean "...related to your organization"?

> In each use case, there are three different classifications of networks. The classifications describe the networks relation to your organization and the network Flex Gateway is in:

Reviewer: ```suggestion
In each use case, there are three different classifications of networks. The classifications describe the relation of each network to your organization and the network that Flex Gateway is in:
```

* Organization internal
+

> Networks owned by your organization and in the same network as Flex Gateway. 

Reviewer: Remove periods from the list items.

* Organization external
+
Trusted networks external to your organization but in a different network than Flex Gateway, for example, your organization’s Splunk instance. 
* Non-Organization external
+

> External networks untrusted by your company and in a different network as Flex Gateway. An API provider’s or API consumers network status affects the necessary security protocols (mTLS) you take when communicating between networks.

Reviewer: ```suggestion
External networks untrusted by your company and in a different network as Flex Gateway. An API provider’s or API consumer's network status affects the necessary security protocols (mTLS) to use when communicating between networks.
```

> If an API provider or API consumer is organization-owned, it belongs to either the organization's internal network or the organization's external networks.

Reviewer: ```suggestion
If an API provider or API consumer is organization-owned, it belongs to either the organization's internal network or its external networks.
```

The following tables:

> * Relates the previously mentioned use cases to use case patterns

Reviewer: ```suggestion
* Relate the previously mentioned use cases to use case patterns
```

> * Shows the networks involved in the pattern

Reviewer: ```suggestion
* Show the networks involved in the pattern
```

> * Indicates whether Flex Gateway acts as an ingress controller, egress contoller, or both.

Reviewer: ```suggestion
* Indicate whether Flex Gateway acts as an ingress controller, egress controller, or both
```

== UC1: Organization-Owned API Exposed to an Organization-Owned API Consumer

[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Pattern | Provider Network | Consumer Network | Role of Flex Gateway

> | P1A: Organization internal API exposed to an organization internal consumer using the same VPC or network as the API | Organization internal | Organization internal | Ingress

Reviewer: ```suggestion
| P1A: Organization internal API exposed to an organization internal consumer that uses the same VPC or network as the API | Organization internal | Organization internal | Ingress
```

> | P1B: Organization internal API exposed to an organization internal consumer using a different VPC or network from the API | Organization internal | Organization internal | Ingress

Reviewer: ```suggestion
| P1B: Organization internal API exposed to an organization internal consumer that uses a different VPC or network from the API | Organization internal | Organization internal | Ingress
```

| P2: Organization internal API exposed to an organization external consumer | Organization internal | Organization external | Ingress
| P3: Organization external API exposed to an organization internal consumer | Organization external | Organization internal | Egress

> | P4: Organization external API exposed to an organization external consumer | Organization external | Organization external | Ingress and Egress

Reviewer: ```suggestion
| P4: Organization external API exposed to an organization external consumer | Organization external | Organization external | Ingress and egress
```

|===

== UC2: Organization-Owned API Exposed to a Non-Organization-Owned API Consumer 

[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Pattern | Provider Network | Consumer Network | Role of Flex Gateway
| P5: Organization internal API exposed to a non-organization external consumer | Organization internal | Non-organization external | Ingress

> | P6: Organization external API exposed to a non-organization external consumer | Organization external | Non-organization external | Ingress and Egress

Reviewer: ```suggestion
| P6: Organization external API exposed to a non-organization external consumer | Organization external | Non-organization external | Ingress and egress
```

|===

== UC3: Non-Organization-Owned API Exposed to an Organization-Owned API Consumer 

[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Pattern | Provider Network | Consumer Network | Role of Flex Gateway
| P7: Non-organization external API exposed to an organization internal consumer | Non-organization external | Organization internal | Egress

> | P8: Non-organization external API exposed to an organization external consumer | Non-organization external | Organization external | Ingress and Egress

Reviewer: ```suggestion
| P8: Non-organization external API exposed to an organization external consumer | Non-organization external | Organization external | Ingress and egress
```

|===

== UC4: Non-Organization-Owned API Exposed to an External Non-Organization-Owned API Consumer 
[%header%autowidth.spread,cols="a,a,a,a"]
|===
| Pattern | Provider Network | Consumer Network | Role of Flex Gateway

> | P9: Non-organization external API exposed to a non-organization external consumer | Non-organization external | Non-organization external | Ingress and Egress

Reviewer: ```suggestion
| P9: Non-organization external API exposed to a non-organization external consumer | Non-organization external | Non-organization external | Ingress and egress
```

=== Certificate authorities
Different certificate authorities (CA) participate in the mTLS connection, depending on the type of integration. In the use case diagrams, color codes and labels represent the different mTLS contexts. Each mTLS context contains a truststore that contains CAs for validating far-side certificates and a keystore that contains a certificate issued by this same CA. The following table describes the three possible mTLS contexts:

[%header%autowidth.spread,cols="a,a,a"]
|===
| mTLS context symbol | Certificate authority | Description
| image:ad-internal-org-mtls.png[white hexagon with mtls written in the middle] | Organization internal mTLS | For internal network communication, the expectation is that only the customer organization uses these certificates within the internal network.
| image:ad-external-org-mtls.png[green hexagon with mtls written in the middle] | Organization external mTLS | For external communication with trusted parties over the public internet, the organization can issue certificates that external consumers or third parties can use to access the services provided by the organization.

> | image:ad-external-non-org-mtls.png[grey hexagon with mtls written in the middle] | Non-organization external mTLS | For external communication with entities that aren't trusted, the third party organization controls the CA. The third party CA issues certificates for the organization to use when communicating with the third-party services.

Reviewer: ```suggestion
| image:ad-external-non-org-mtls.png[grey hexagon with mtls written in the middle] | Non-organization external mTLS | For external communication with entities that aren't trusted, the third-party organization controls the CA. The CA issues certificates for the organization to use when communicating with third-party services.
```

= Integration Use Cases
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> The following integration use cases are determined by the ownership of the API Provider and the ownership of the API Consumer. Each use case also provides physical patterns demonstrating the necessary services in the system. 

Reviewer: ```suggestion
The following integration use cases are determined by the ownership of the API Provider and API Consumer. Each use case shows physical patterns that demonstrate the necessary services in the system. 
```

> For all use case diagrams, Flex Gateway is running in Connected Mode. However, you can also run Flex Gateway in Local Mode for all use cases. You can also use a combination of Connected Mode and Local Mode for separate runtimes in the same or in different environments.

Reviewer: ```suggestion
For the following use case diagrams, Flex Gateway is running in Connected Mode. However, you can also run Flex Gateway in Local Mode or use a combination of Connected Mode and Local Mode for separate runtimes in the same or in different environments.
```

> NOTE: The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks with the appropriate network controls applied.

Reviewer: I suggest removing the NOTE:

> NOTE: The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks with the appropriate network controls applied.

Reviewer: ```suggestion
The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks that have the appropriate network controls applied.
```

> NOTE: The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks with the appropriate network controls applied.

Reviewer: ```suggestion
The use case diagrams use Kubernetes terminology, but you can extend the architecture to other technology stacks that have the appropriate network controls applied.
```

Flex Gateway supports four use cases:

> * xref:flex-architecture-uc1.adoc[]

Reviewer: Please check that the link text renders OK.

* xref:flex-architecture-uc2.adoc[] 
* xref:flex-architecture-uc3.adoc[] 
* xref:flex-architecture-uc4.adoc[] 

> image:ad-intergration-use-cases.png[The four use cases are linked to products they could support.]

Reviewer: ```suggestion
image:ad-intergration-use-cases.png[The four use cases are linked to products that they can support.]
```

> Each use case supports different integration patterns that vary depending on the location of APIs relative to the network Flex Gateway is in. You can use different use case patterns depending on the different API owners and networks in your system. 

Reviewer: ```suggestion
Each use case supports different integration patterns that vary depending on the location of APIs relative to the network in which Flex Gateway is located. You can use different use case patterns, depending on the different API owners and networks in your system. 
```

In each use case, there are three different classifications of networks. The classifications describe the networks relation to your organization and the network Flex Gateway is in:

> * Organization internal: networks owned by your organization and in the same network as Flex Gateway. 

Reviewer: ```suggestion
* Organization internal
+
Networks owned by your organization and in the same network as Flex Gateway
```

Reviewer: When using a list, it's best to add a blank line after a term before you describe it. It makes the term stand out. I didn't mark this everywhere.

* Organization external: trusted networks external to your organization but in a different network than Flex Gateway, for example, your organization’s Splunk instance. 

> * Non-Organization external: networks not trusted by your company and not in the same network as Flex Gateway. An API provider’s or API consumers network status affects the necessary security protocols (mTLS) you take when communicating.

Reviewer: ```suggestion
* Non-Organization external
+
External networks untrusted by your company and in a different network than Flex Gateway. An API provider’s or API consumer's network status affects the required security protocols (mTLS) to use when communicating.
```

Reviewer: when communicating in these networks?

> If an API Provider or API Consumer is Organization owned, it belongs to either organization internal or organization external networks.

Reviewer: ```suggestion
If an API Provider or API Consumer is organization-owned, it belongs to either the organization's internal network or its external networks.
```

> If an API Provider or API Consumer is Organization owned, it belongs to either organization internal or organization external networks.

Reviewer: Do the words Provider, Consumer and Organization need to be initial capped?

> The following table relates the previously mentioned use cases to use case patterns, the networks involved in the pattern, and whether Flex Gateway acts as an ingress or egress contoller or both.  

Reviewer: Suggestion:

The following table does the following:

* Relates the documented use cases to use case patterns
* Shows the networks used in the patterns
* Indicates whether Flex Gateway acts as an ingress controller, egress controller, or both

[%header%autowidth.spread,cols="a,a,a,a,a"]
|===

> | Use Case | Pattern | Provider Network | Consumer Network | Role of Flex Gateway

Reviewer: In this table, once you mention a use case in the first column, the cell should be blank for the other occurrences of that use case. Ideally, the cells in the first column for the same use case should be merged. 

I tried to show an example here but git messes up the formatting.

Reviewer: If you want to simplify the table, you could have a separate heading for each use and then have a table that shows only that use case. I think this might be a better way to go.

== UC1:  Organization-Owned API Exposed to an Organization-Owned API Consumer

Pattern    Provider Network    Consumer Network    Role of Flex Gateway

=== UC2: 
Pattern     Provider Network     Consumer Network   Role of Flex Gateway

> | UC1: Organization Owned API Exposed to Organization Owned API Consumer | P1A: Organization internal API exposed to organization internal consumer using the same VPC or network | Organization internal | Organization internal | Ingress

Reviewer: ```suggestion
| UC1: Organization-Owned API Exposed to Organization Owned API Consumer | P1A: Organization internal API exposed to organization internal consumer using the same VPC or network | Organization internal | Organization internal | Ingress
```

Reviewer: Global: I would initial-cap the use cases instead of using title caps.
Also, organization owned as an adjective needs to be hyphenated, for example, "organization-owned API." I won't mark all of these.

"exposed to organization internal consumer" > "exposed to an organization's internal consumer"

> | UC1: Organization Owned API Exposed to Organization Owned API Consumer | P1B: Organization internal API exposed to organization internal consumer using a different VPC or network | Organization internal | Organization internal | Ingress

Reviewer: ```suggestion
| UC1: Organization-owned API exposed to an organization-owned API Consumer | P1B: Organization internal API exposed to an organization internal consumer that uses a different VPC or network | Organization internal | Organization internal | Ingress
```

Reviewer: ..."an organization internal consumer that uses a different VPC or network"

Different than what? Different from the one the organization uses?

| UC2: Organization Owned API Exposed to Non-Organization owned API Consumer | P6: Organization external API exposed to non-organization external consumer | Organization external | Non-organization external | Ingress and Egress
| UC3: Non-Organization Owned API Exposed to Organization Owned API Consumer | P7: Non-organization external API exposed to organization internal consumer | Non-organization external | Organization internal | Egress
| UC3: Non-Organization Owned API Exposed to Organization Owned API Consumer | P8: Non-organization external API exposed to Organization external consumer | Non-organization external | Organization external | Ingress and Egress
| UC4: Non-Organization owned API Exposed to External Non-Organization owned API Consumer | P9: Non-organization external API exposed to non-organization external consumer | Non-organization external | Non-organization external | Ingress and Egress
|===

== mTLS considerations for integration flows
Flex Gateway supports Mutual Transport Layer Security (mTLS) for all Flex Gateway initiated or terminated connections. Mulesoft reconmends using mTLS on all communication flows. 

> For inbound connections that cross a network security boundary, Mulesoft recommends using a network application firewall as close to the boundary as possible. For inbound connections, the user configures the external mTLS connection to terminate at the firewall and configures a new mTLS context between the firewall and Flex Gateway.

Reviewer: ```suggestion
For inbound connections that cross a network security boundary, use a network application firewall as close to the boundary as possible. For inbound connections, configure the external mTLS connection to terminate at the firewall and configure a new mTLS context between the firewall and Flex Gateway.
```

Reviewer: We try to avoid "MuleSoft recommends" per our style guide.

=== Certificate authorities

> Different certificate authorities (CA) participate in the mTLS connection depending on the type of integration. In the use cases, we consider three CA classes. Color codes and labels represent the different mTLS contexts. Each mTLS context contains a truststore containing CAs to validate far-side certificates and certificates and a keystore containing a certificate issued by this same CA. The following table describes the three possible mTLS contexts.

Reviewer: ```suggestion
Different certificate authorities (CA) participate in the mTLS connection, depending on the type of integration. In the use case diagrams, color codes and labels represent the different mTLS contexts. Each mTLS context contains a truststore that contains CAs for validating far-side certificates and a keystore that contains a certificate issued by this same CA. The following table describes the three possible mTLS contexts:
```

[%header%autowidth.spread,cols="a,a,a"]
|===
| mTLS context symbol | Certificate authority | Description
| image:ad-internal-org-mtls.png[white hexagon with mtls written in the middle] | Organization internal mTLS | For internal network communication, the expectation is that only the customer organization uses these certificates within the internal network.

> | image:ad-external-org-mtls.png[green hexagon with mtls written in the middle] | Organization external mTLS | For external communication with trusted parties over the public internet, the organization may issue certificates using this CA for use by external consumers first or third parties to access services provided by the organization.

Reviewer: ```suggestion
| image:ad-external-org-mtls.png[green hexagon with mtls written in the middle] | Organization external mTLS | For external communication with trusted parties over the public internet, the organization can issue certificates that external consumers or third parties can use to access the services provided by the organization.
```

> | image:ad-external-non-org-mtls.png[grey hexagon with mtls written in the middle] | Non-organization external mTLS | For external communication with entities that aren't trusted, the third party organization controls the CA. The third party CA issues certificates to the organization to use when communicating with the third party services.

Reviewer: ```suggestion
| image:ad-external-non-org-mtls.png[grey hexagon with mtls written in the middle] | Non-organization external mTLS | For external communication with entities that aren't trusted, the third party organization controls the CA. The third party CA issues certificates for the organization to use when communicating with the third-party services.
```
